pinot-core/src/main/java/com/linkedin/pinot/core/realtime/converter/RealtimeSegmentConverter.java

```diff
@@ -150,4 +137,33 @@ public void build(@Nullable SegmentVersion segmentVersion, ServerMetrics serverM
       }
     }
   }
+
+  /**
+   * Returns a new schema based on the original one. The new schema removes columns as needed (for ex, virtual cols)
+   * and adds the new timespec to the schema.
+   */
+  @VisibleForTesting
+  public
+  Schema getUpdatedSchema(Schema original) {
+
+    TimeFieldSpec tfs = original.getTimeFieldSpec();
+    // Use outgoing granularity for creating segment
+    TimeGranularitySpec outgoing = tfs.getOutgoingGranularitySpec();
+    TimeFieldSpec newTimeSpec = new TimeFieldSpec(outgoing);
+
+    Schema newSchema = new Schema();
```

Comment:
Will iterating over schema.getPhysicalColumns() work here? (since you added that method)

Reply:
Good point. Updated.